### DISCUSSION TITLE
Does compiled queries support nullables and enums?

### DISCUSSION THREAD

**Discussion by ushenkodmitry** [JasperFx/marten#2714] [Q&A]
Hello folks!
My compiled query is defined like so:
```
public sealed record DeliveryQuery : ICompiledListQuery<DeliveryEntity>
{
	public int? Id{ get; init; }

	public DeliveryProviderName[] ProviderNames { get; init; }; // enum

	public Expression<Func<IMartenQueryable<DeliveryEntity>, IEnumerable<DeliveryEntity>>> QueryIs()
	{
		return q => q.Where(e => e.ProviderName.IsOneOf(ProviderNames) && e.KitchenId == KitchenId);
	}
}
```

When executing i'm getting exceptions:
1. Could not determine an 'id/Id' field or property for requested document type DeliveryProviderName (which is enum)
2. Members Nullable{int} Id cannot be used as parameters to a compiled query 

Can I use these types in compiled queries?

**Comment by mysticmind** [comment]
https://martendb.io/documents/identity.html - Marten looks for an id/Id/ID property of type int/string/guid etc (refer to the document). If you want to use a different id property then you could use  `Identity` attribute. These same rules apply when you define a compiled query. You can't use a nullable type for identity.

**Comment by ushenkodmitry** [comment]
@mysticmind thanks for the answer. But I think you didn't understand me. 
I have a document **DeliveryEntity**
```
public sealed record DeliveryEntity
{
public Guid CorrelationId { get; set; }
public DeliveryProviderName ProviderName { get; set; }
public int KitchenId { get; set; }
...
}
```
And I have a compiled query
```
public sealed record DeliveriesQuery : ICompiledListQuery<DeliveryEntity>
{
        public int? KitchenId {get;init;}
	public DeliveryProviderName[] ProviderNames { get; init; } = Array.Empty<DeliveryProviderName>();

	public Expression<Func<IMartenQueryable<DeliverySagaEntity>, IEnumerable<DeliverySagaEntity>>> QueryIs()
	{
		if (ProviderNames.Any())
			return q => q.Where(e => ProviderNames.Contains(e.ProviderName));

		return q => q;
	}
}
```
**CorrelationId** property configured as document identity for DeliveryEntity. There is no problem with that. 
The problem is: "Could not determine an 'id/Id' field or property for requested document type **DeliveryProviderName**." But it is a property on **DeliveryEntity** document of type **DeliveryProviderName**. Which is enum. 
`public enum DeliveryProviderName { None, SM }`

What kind of id it wants for enum?

PS The query works fine if I remove the properties altogether from it.

**Reply by mysticmind** [reply]
Okay, heading to bed, will check with a fresh pair of eyes tomorrow and revert to you.

**Comment by ushenkodmitry** [comment]
Thank you very much.

пн, 11 сент. 2023 г., 19:35 Babu Annamalai ***@***.***>:

**Comment by ushenkodmitry** [comment]
@mysticmind
Hello,
did you have a chance to have a fresh pair of eyes? ) It is also unclear if nullables can be used as compiled query params. Having nullable properties in compiled query resulting in InvalidCompiledQueryException: Members Nullable<System.DateTime> UpdatedFrom, Nullable<System.DateTime> UpdatedTo cannot be used as parameters to a compiled query
